File: src/agent/python/services.py
import asyncio
import time
import uuid
from datetime import timedelta
from logging import Logger
from typing import Any, Coroutine, Optional, cast

from apollo3zehn_package_management import ExtensionHive, PackageService
from nexus_extensibility import IDataSource
from nexus_remoting._remoting import RemoteCommunicator


class TcpClientPair:
    comm_reader: Optional[asyncio.StreamReader] = None
    comm_writer: Optional[asyncio.StreamWriter] = None
    data_reader: Optional[asyncio.StreamReader] = None
    data_writer: Optional[asyncio.StreamWriter] = None
    remote_communicator: Optional[RemoteCommunicator] = None
    watchdog_timer = time.time()
    task: Optional[asyncio.Task] = None

class AgentService:

    CLIENT_TIMEOUT = timedelta(minutes=1)

    _background_tasks = set[asyncio.Task]()
    _tcp_client_pairs: dict[uuid.UUID, TcpClientPair] = {}
    _lock = asyncio.Lock()

    def __init__(
            self, 
            extension_hive: ExtensionHive, 
            package_service: PackageService, 
            logger: Logger, 
            json_rpc_listen_address: str,
            json_rpc_listen_port: int
        ):
        
        self._extension_hive = extension_hive
        self._package_service = package_service
        self._logger = logger
        self._json_rpc_listen_address = json_rpc_listen_address
        self._json_rpc_listen_port = json_rpc_listen_port

    async def load_packages(self):

        self._logger.info("Load packages")
        package_reference_map = await self._package_service.get_all()
        await self._extension_hive.load_packages(package_reference_map)

    async def accept_clients(self):

        async def detect_and_remove_inactive_clients():

            while True:

                await asyncio.sleep(600)

                for key, pair in list(self._tcp_client_pairs.items()):

                    now = time.time()
                    watchdog_timer_elasped = timedelta(seconds=now - pair.watchdog_timer)

                    is_dead =\
                        (pair.comm_reader is None or pair.data_reader is None) and watchdog_timer_elasped >= self.CLIENT_TIMEOUT or \
                        pair.remote_communicator is not None and pair.remote_communicator.last_communication >= self.CLIENT_TIMEOUT

                    if is_dead:

                        if key in self._tcp_client_pairs:

                            # TODO: Add proper cancellation https://medium.com/nuculabs/cancellation-token-pattern-in-python-b549d894e244
                            # Otherwise sockets may stay open
                            if pair.task is not None:
                                pair.task.cancel()

                            del self._tcp_client_pairs[key]

        self._create_task(detect_and_remove_inactive_clients())
        
        server = await asyncio.start_server(
            self._handle_client, 
            host=self._json_rpc_listen_address,
            port=self._json_rpc_listen_port
        )

        self._logger.info(
            "Listening for JSON-RPC communication on %s:%d",
            self._json_rpc_listen_address,
            self._json_rpc_listen_port
        )

        async with server:
            await server.serve_forever()

    async def _handle_client(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):

        # Get connection id
        buffer1 = await asyncio.wait_for(reader.readexactly(36), timeout=5)
        id_string = buffer1.decode("utf-8")

        # Get connection type
        buffer2 = await asyncio.wait_for(reader.readexactly(4), timeout=5)
        type_string = buffer2.decode("utf-8")

        try:
            id = uuid.UUID(id_string)

        except:
            writer.close()
            await writer.wait_closed()
            return

        self._logger.debug("Accept TCP client with connection ID %s and communication type %s", id_string, type_string)

        async with self._lock:

            # We got a "comm" tcp connection
            if type_string == "comm":

                if id not in self._tcp_client_pairs:
                    self._tcp_client_pairs[id] = TcpClientPair()

                self._tcp_client_pairs[id].comm_reader = reader
                self._tcp_client_pairs[id].comm_writer = writer

            # We got a "data" tcp connection
            elif type_string == "data":

                if id not in self._tcp_client_pairs:
                    self._tcp_client_pairs[id] = TcpClientPair()

                self._tcp_client_pairs[id].data_reader = reader
                self._tcp_client_pairs[id].data_writer = writer
                
            # Something went wrong, close the socket and return
            else:
                writer.close()
                await writer.wait_closed()
                return

            pair = self._tcp_client_pairs[id]

            if pair.comm_reader and \
               pair.comm_writer and \
               pair.data_reader and \
               pair.data_writer and \
               not pair.remote_communicator:

                self._logger.debug("Accept remoting client with connection ID %s", id)

                pair.remote_communicator = RemoteCommunicator(
                    pair.comm_reader,
                    pair.comm_writer,
                    pair.data_reader,
                    pair.data_writer,
                    get_data_source_type=lambda type_name: self._extension_hive.get_extension_type(type_name)
                )

                pair.task = self._create_task(pair.remote_communicator.run())

    def _create_task(self, coro: Coroutine[Any, Any, Any]) -> asyncio.Task:

        task = asyncio.create_task(coro)
        self._background_tasks.add(task)
        task.add_done_callback(self._background_tasks.discard)

        return task

File: src/agent/rust/src/services.rs
use log::{debug, info};
use nexus_remoting::RemoteCommunicator;
use smol::io::AsyncReadExt;
use smol::net::{TcpListener, TcpStream};
use smol::{spawn, Task, Timer};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};
use uuid::Uuid;

struct TcpClientPair {
    comm_stream: Option<TcpStream>,
    data_stream: Option<TcpStream>,
    remote_communicator: Option<RemoteCommunicator>,
    watchdog_timer: Instant,
    task: Option<Task<()>>
}

impl TcpClientPair {
    fn new() -> Self {
        Self {
            comm_stream: None,
            data_stream: None,
            remote_communicator: None,
            watchdog_timer: Instant::now(),
            task: None
        }
    }
}

pub struct AgentService {
    extension_hive: ExtensionHive,
    tcp_client_pairs: Arc<Mutex<HashMap<Uuid, TcpClientPair>>>,
    json_rpc_listen_address: String,
    json_rpc_listen_port: u16,
    client_timeout: Duration
}

impl AgentService {
    pub fn new(
        extension_hive: ExtensionHive,
        json_rpc_listen_address: String,
        json_rpc_listen_port: u16,
    ) -> Self {
        Self {
            extension_hive,
            tcp_client_pairs: Arc::new(Mutex::new(HashMap::new())),
            json_rpc_listen_address,
            json_rpc_listen_port,
            client_timeout: Duration::from_secs(60)
        }
    }

    pub async fn accept_clients(&self) -> smol::io::Result<()> {
        let tcp_client_pairs = self.tcp_client_pairs.clone();
        let client_timeout = self.client_timeout;

        // Spawn a task to detect and remove inactive clients
        spawn(async move {
            loop {
                Timer::after(Duration::from_secs(600)).await;

                let mut pairs = tcp_client_pairs.lock().unwrap();
                let now = Instant::now();

                pairs.retain(|_, pair| {
                    let watchdog_timer_elapsed = now.duration_since(pair.watchdog_timer);

                    let is_dead = 
                        (pair.comm_stream.is_none() || pair.data_stream.is_none()) && watchdog_timer_elapsed >= client_timeout ||
                        pair.remote_communicator.as_ref().is_some_and(|x| x.last_communication() >= client_timeout);

                    if is_dead {
                        if let Some(task) = pair.task.take() {
                            task.cancel();
                        }
                    }

                    !is_dead
                });
            }
        }).detach();

        // Start the TCP server
        let listener = TcpListener::bind((&self.json_rpc_listen_address[..], self.json_rpc_listen_port)).await?;

        info!(
            "Listening for JSON-RPC communication on {}:{}",
            self.json_rpc_listen_address, self.json_rpc_listen_port
        );

        loop {
            let (stream, _) = listener.accept().await?;
            let tcp_client_pairs = self.tcp_client_pairs.clone();

            smol::spawn(async move {
                if let Err(e) = self.handle_client(stream, tcp_client_pairs).await {
                    eprintln!("Error handling client: {}", e);
                }
            })
            .detach();
        }
    }

    async fn handle_client(
        &self,
        stream: TcpStream,
        tcp_client_pairs: Arc<Mutex<HashMap<Uuid, TcpClientPair>>>,
    ) -> smol::io::Result<()> {
        let mut reader = stream.clone();
        let writer = stream;

        // Get connection id
        let mut buffer = vec![0u8; 36];

        reader.read_exact(&mut buffer).await?;

        let id_string = String::from_utf8(buffer).unwrap();
        let id = Uuid::parse_str(&id_string).unwrap();

        // Read connection type
        let mut buffer = vec![0u8; 4];

        reader.read_exact(&mut buffer).await?;

        let type_string = String::from_utf8(buffer).unwrap();

        debug!("Accept TCP client with connection ID {id_string} and communication type {type_string}");

        let mut pairs = tcp_client_pairs.lock().unwrap();
        let pair = pairs.entry(id).or_insert_with(TcpClientPair::new);

        match type_string.as_str() {
            "comm" => {
                pair.comm_stream = Some(writer);
            }
            "data" => {
                pair.data_stream = Some(writer);
            }
            _ => {
                // todo!("Close the socket!! (writer.close())")
                return Err(smol::io::Error::new(
                    smol::io::ErrorKind::InvalidData,
                    "Unknown connection type",
                ));
            }
        }

        if pair.comm_stream.is_some() && pair.data_stream.is_some() && pair.remote_communicator.is_none() {
            debug!("Accept remoting client with connection ID {id}");

            let comm_stream = pair.comm_stream.take().unwrap();
            let data_stream = pair.data_stream.take().unwrap();
            let get_data_source_type = |type_name: &str| self.extension_hive.get_extension_type(type_name);

            pair.remote_communicator = Some(RemoteCommunicator::new(comm_stream, data_stream, get_data_source_type));
            pair.remote_communicator.as_ref().unwrap().run().await?;
        }

        Ok(())
    }
}